cpp/dolfinx/common/IndexMap.cpp

```diff
@@ -241,19 +248,60 @@ compute_submap_indices(const IndexMap& imap,
     std::ranges::sort(global_idx_to_possible_owner);
 
     // Choose the submap owner for each index in `recv_indices`
+    // and pack destination ranks for each process that has received new indices
     std::vector<int> send_owners;
     send_owners.reserve(recv_indices.size());
-    for (auto idx : recv_indices)
+    std::vector<int> new_owner_dest_ranks;
+    new_owner_dest_ranks.reserve(recv_indices.size());
+    std::vector<int> new_owner_dest_ranks_offsets(recv_sizes.size() + 1, 0);
+    std::vector<std::int32_t> new_owner_dest_ranks_sizes(recv_sizes.size());
+    new_owner_dest_ranks_sizes.reserve(1);
+
+    for (std::size_t i = 0; i < recv_sizes.size(); ++i)
     {
-      // NOTE: Could choose new owner in a way that is is better for
-      // load balancing, though the impact is probably only very small
-      auto it = std::ranges::lower_bound(global_idx_to_possible_owner, idx,
-                                         std::ranges::less(),
-                                         [](auto& e) { return e.first; });
-      assert(it != global_idx_to_possible_owner.end() and it->first == idx);
-      send_owners.push_back(it->second);
-    }
+      for (int j = recv_disp[i]; j < recv_disp[i + 1]; ++j)
+      {
+        std::int64_t idx = recv_indices[j];
+
+        // NOTE: Could choose new owner in a way that is is better for
+        // load balancing, though the impact is probably only very small
+        auto it = std::ranges::lower_bound(global_idx_to_possible_owner, idx,
+                                           std::ranges::less(),
+                                           [](auto& e) { return e.first; });
+        assert(it != global_idx_to_possible_owner.end() and it->first == idx);
+        send_owners.push_back(it->second);
 
+        // If rank that sent this ghost is the owner, send all other ranks
+        if (it->second == dest[i])
+        {
+          // Find upper limit of recv index and pack all dest ranks for this
+          // process
+          auto it_upper = std::ranges::upper_bound(
+              it, global_idx_to_possible_owner.end(), idx, std::ranges::less(),
+              [](auto& e) { return e.first; });
+          std::transform(it + 1, it_upper,
+                         std::back_inserter(new_owner_dest_ranks),
+                         [](auto& e) { return e.second; });
+        }
+      }
+      // Remove duplicate new dest ranks from recv process
```

Comment:
Sorry, can you explain why there would be duplicates? Since `new_owner_dest_ranks` is created from `global_idx_to_possible_owner`, do we not already have a unique set of ranks for each index?

Reply:
We only need the unique set of ranks for all newly assigned indices to the process.
I.e. if a process has taken over index i from process (N, M) and index j from (M, N, Q), what we want to send to this process is (M,N, Q), not (N,M,M,N,Q), as one would then have to do a sort and remove on all received indices.